### ISSUE TITLE
torch.unique performance issue for return_inverse=True

### ISSUE THREAD

**Issue by ptrblck** [pytorch/pytorch#15804]
## `torch.unique` performance issues for large tensors with a large number of unique values

For large tensors (`1e7` values in one dimension) `torch.unique` performs badly, if `return_inverse=True` is set.
Since the code is currently split into the "non-dim" and "dim" version, we should take a look at them separately.

The issue was reported [here](https://discuss.pytorch.org/t/torch-unique-is-very-slow-for-big-number-of-uniques/30648) the first time by Etienne Perot.

## To Reproduce

I used the following script to time the code:
```python
import argparse
import time
import torch

def test_unique(use_cuda, use_dim):
    N = int(1e7)
    device = 'cuda:0' if use_cuda else 'cpu'
    for high in torch.logspace(1, 5, 5):
        high = int(high.item())
        print('Timing unique with high={} for {} values on {}'.format(
            high, N, device))
        x = torch.randint(low=0, high=high, size=(N,))
        if use_cuda:
            x = x.to(device)

        torch.cuda.synchronize()
        start = time.time()
        if use_dim:
            unique, inv = torch.unique(
                x, sorted=False, return_inverse=True, dim=0)
        else:
            unique, inv = torch.unique(x, sorted=False, return_inverse=True)
        torch.cuda.synchronize()
        print(time.time()-start)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--use_cuda', action='store_true', default=False)
    parser.add_argument('--use_dim', action='store_true', default=False)
    args = parser.parse_args()
    use_cuda = args.use_cuda and torch.cuda.is_available()
    use_dim = args.use_dim
    test_unique(use_cuda=use_cuda, use_dim=use_dim)


if __name__=='__main__':
    main()
```

We run the `torch.unique` call an a tensor containing `1e7` values with an increasing number of unique values in each iteration.
Here are the timings on my machine (in seconds):
```python
# CPU, non-dim
Timing unique with high=10 for 10000000 values on cpu
0.2487781047821045
Timing unique with high=100 for 10000000 values on cpu
0.25162458419799805
Timing unique with high=1000 for 10000000 values on cpu
0.2525928020477295
Timing unique with high=10000 for 10000000 values on cpu
0.29808735847473145
Timing unique with high=100000 for 10000000 values on cpu
0.37146925926208496

# CPU, use_dim
Timing unique with high=10 for 10000000 values on cpu
14.624972820281982
Timing unique with high=100 for 10000000 values on cpu
15.274183988571167
Timing unique with high=1000 for 10000000 values on cpu
15.500507354736328
Timing unique with high=10000 for 10000000 values on cpu
15.738206148147583
Timing unique with high=100000 for 10000000 values on cpu
15.88893437385559

# GPU, non-dim
Timing unique with high=10 for 10000000 values on cuda:0
0.015486717224121094
Timing unique with high=100 for 10000000 values on cuda:0
0.11904716491699219
Timing unique with high=1000 for 10000000 values on cuda:0
0.41509509086608887
Timing unique with high=10000 for 10000000 values on cuda:0
6.064100980758667
Timing unique with high=100000 for 10000000 values on cuda:0
63.65047788619995

# GPU, use_dim
Timing unique with high=10 for 10000000 values on cuda:0
371.14407658576965
Timing unique with high=100 for 10000000 values on cuda:0
351.54057216644287
# Probably no need to further time here
```

While the CPU-non-dim implementation seems reasonable, all others show some performance issues, especially the dim-implementations.

I've used [these tests](https://gist.github.com/ptrblck/ac0dc862f4e1766f0e1036c252cdb105) to time the performance while adding the `dim` argument [here](https://github.com/pytorch/pytorch/pull/10423).
Apparently these tests did not cover enough use cases, so I'm sorry about it.

## Debugging

It looks like the non-dim GPU implementation is using a nested loop as the CUDA kernel [here](https://github.com/pytorch/pytorch/blob/b6a8c45f57b65d11894c4a6e5a3267708ecec1c5/aten/src/ATen/native/cuda/Unique.cu#L25) which depends on the number of inputs and outputs.

After timing the dim implementation using `AT_CUDA_CHECK(cudaStreamSynchronize(stream));` and a `std::clock`, it seems the bottleneck for the dim-implementation is in calculating the inverse indices [here](https://github.com/pytorch/pytorch/blob/b6a8c45f57b65d11894c4a6e5a3267708ecec1c5/aten/src/ATen/native/cuda/Unique.cu#L133).

I'm currently timing the CPU-dim implementation to narrow down the lines of code responsible for this issue.

While our last tests have shown a reasonable performance for all test cases, some edge cases were missed. I won't be able to work on this issue in the next three weeks, but would like to have a look at it after. 
This issue might be a blocker for the [return_counts PR](https://github.com/pytorch/pytorch/pull/15495).

Suggestions are welcome. If this needs a proper refactoring, we could also consider combining the non-dim/dim implementations.

CC @colesbury, @ezyang 

Best,
ptrblck

## Environment

PyTorch version: 1.0.0a0+d68edb6
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.12.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 410.78
cuDNN version: Probably one of the following:
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7.4.1
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn_static.a
/usr/local/cuda-9.1/lib64/libcudnn.so
/usr/local/cuda-9.1/lib64/libcudnn.so.7
/usr/local/cuda-9.1/lib64/libcudnn.so.7.4.1
/usr/local/cuda-9.1/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip] Could not collect
[conda] blas                      1.0                         mkl  
[conda] magma-cuda100             2.4.0                         1    pytorch
[conda] mkl                       2019.1                      144  
[conda] mkl-include               2019.1                      144  
[conda] mkl_fft                   1.0.6            py37hd81dba3_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] torch                     1.0.0a0+d68edb6           <pip>

**Comment by jcjohnson** [comment]
I'm running into the same performance issue, and it's currently bottlenecking my training pipeline -- I need to run unique with inverse with ~500k values, which takes tens of seconds on GPU and hundreds of milliseconds on CPU.

Is there any plan for a fix in the near term?

Here's my own benchmarking script:

https://gist.github.com/jcjohnson/2e0b817e9694dc6ee4d40c2a4ab31172

With a Tesla M40 I get the following times:
```
Running with N = 10000, M = 10000
cuda (no inverse): 0.99 ms
cpu (no inverse): 3.95 ms
cuda (with inverse): 2.16 ms
cpu (with inverse): 8.33 ms

Running with N = 10000, M = 100000
cuda (no inverse): 0.68 ms
cpu (no inverse): 5.54 ms
cuda (with inverse): 2.91 ms
cpu (with inverse): 11.66 ms

Running with N = 100000, M = 100000
cuda (no inverse): 1.08 ms
cpu (no inverse): 44.11 ms
cuda (with inverse): 348.97 ms
cpu (with inverse): 93.60 ms

Running with N = 100000, M = 1000000
cuda (no inverse): 0.69 ms
cpu (no inverse): 64.49 ms
cuda (with inverse): 729.19 ms
cpu (with inverse): 145.16 ms

Running with N = 500000, M = 500000
cuda (no inverse): 1.62 ms
cpu (no inverse): 257.77 ms
cuda (with inverse): 19069.52 ms
cpu (with inverse): 508.64 ms

Running with N = 500000, M = 5000000
cuda (no inverse): 1.30 ms
cpu (no inverse): 370.29 ms
cuda (with inverse): 29015.42 ms
cpu (with inverse): 715.31 ms
```

And on a Tesla P100 I get the following:
```
Running with N = 10000, M = 10000
cuda (no inverse): 0.61 ms
cpu (no inverse): 0.88 ms
cuda (with inverse): 1.28 ms
cpu (with inverse): 1.58 ms

Running with N = 10000, M = 100000
cuda (no inverse): 0.42 ms
cpu (no inverse): 1.32 ms
cuda (with inverse): 1.75 ms
cpu (with inverse): 2.54 ms

Running with N = 100000, M = 100000
cuda (no inverse): 0.66 ms
cpu (no inverse): 10.05 ms
cuda (with inverse): 221.86 ms
cpu (with inverse): 19.20 ms

Running with N = 100000, M = 1000000
cuda (no inverse): 0.37 ms
cpu (no inverse): 16.79 ms
cuda (with inverse): 422.35 ms
cpu (with inverse): 31.27 ms

Running with N = 500000, M = 500000
cuda (no inverse): 0.94 ms
cpu (no inverse): 52.24 ms
cuda (with inverse): 11084.27 ms
cpu (with inverse): 106.10 ms

Running with N = 500000, M = 5000000
cuda (no inverse): 0.73 ms
cpu (no inverse): 92.83 ms
cuda (with inverse): 16980.49 ms
cpu (with inverse): 177.26 ms
```

**Comment by ezyang** [comment]
got fixed!
